File: 2421-number-of-good-paths/2421-number-of-good-paths.cpp
class Solution {
public:
    struct Disj{
        vector<int> parent;
        
        Disj(int n){
            parent.resize(n);
            iota(parent.begin(), parent.end(), 0);
        }
        
        int find(int x){
            if(x == parent[x]){
                return x;
            }
            return parent[x] = find(parent[x]);
        }
        
        void join(int x, int y){
            x = find(x);
            y = find(y);
            if(x != y){
                parent[x] = y;
            }
        }
    };
    
    int numberOfGoodPaths(vector<int>& vals, vector<vector<int>>& edges) {
        int n = vals.size(), goodPaths = 0;
        vector<vector<int>> adj(n);
        map<int, vector<int>> sameValues;
        
        for(int i = 0; i < n; i++){
            sameValues[vals[i]].push_back(i);
        }
        
        for(auto &e : edges){
            int u = e[0], v = e[1];
            
            if(vals[u] >= vals[v]){
                adj[u].push_back(v);
            } else {
                adj[v].push_back(u);
            }
        }
        
        Disj d(n);
        
        for(auto &[value, allNodes] : sameValues){
            for(int u : allNodes){
                for(int v : adj[u]){
                    d.join(u, v);
                }
            }
            
            unordered_map<int, int> group;
            
            for(int u : allNodes){
                group[d.find(u)]++;
            }
            
            goodPaths += allNodes.size();
            
            for(auto &[_, size] : group){
                goodPaths += (size * (size - 1) / 2);
            }
        }
        
        return goodPaths;
    }
};

File: 2421-number-of-good-paths/2421-number-of-good-paths.java
class Solution {
    class Disj {
        int[] parent;
        
        Disj(int n){
            parent = new int[n];
            for(int i = 0; i < n; i++){
                parent[i] = i;
            }
        }
        
        int find(int x){
            if(x != parent[x]){
                return parent[x] = find(parent[x]);
            }
            return x;
        }
        
        void join(int x, int y){
            x = find(x);
            y = find(y);
            
            if(x != y){
                parent[x] = y;
            }
        }
    }
    
    public int numberOfGoodPaths(int[] vals, int[][] edges) {
        int n = vals.length, ans = 0;
        List<Integer>[] adj = new List[n];
        Map<Integer, List<Integer>> sameVals = new TreeMap<>();
        
        for(int i = 0; i < n ;i++){
            adj[i] = new ArrayList<>();
            sameVals.computeIfAbsent(vals[i], k -> new ArrayList<>()).add(i);
        }
        
        for(int[] e : edges){
            int u = e[0], v = e[1];
            
            if(vals[u] >= vals[v]){
                adj[u].add(v);
            } else {
                adj[v].add(u);
            }
        }
        
        Disj d = new Disj(n);
        
        for(int val : sameVals.keySet()){
            for(int u : sameVals.get(val)){
                for(int v : adj[u]){
                    d.join(u, v);
                }
            }
            
            Map<Integer, Integer> group = new HashMap<>();
            
            for(int u : sameVals.get(val)){
                group.put(d.find(u), group.getOrDefault(d.find(u), 0) + 1);
            }
            
            ans += sameVals.get(val).size();
            
            for(int key : group.keySet()){
                int size = group.get(key);
                ans += size * (size-1) / 2;
            }
        }
        
        return ans;
    }
}